File: src/lib/estimates.ts
const ESTIMATE_SCALES: Record<string, { base: number[]; extended: number[] }> = {
  exponential: { base: [1, 2, 4, 8, 16], extended: [32, 64] },
  fibonacci: { base: [1, 2, 3, 5, 8, 13], extended: [21, 34] },
  linear: { base: [1, 2, 3, 4, 5], extended: [6, 7, 8, 9, 10] },
  tShirt: { base: [1, 2, 3, 4, 5], extended: [6] },
};

const TSHIRT_LABELS: Record<number, string> = {
  0: "None",
  1: "XS",
  2: "S",
  3: "M",
  4: "L",
  5: "XL",
  6: "XXL",
};

export type EstimateConfig = { type: string; allowZero: boolean; extended: boolean };

/** Build an EstimateConfig from a team object's estimation properties. */
export function buildEstimateConfig(team: {
  issueEstimationType: string;
  issueEstimationAllowZero: boolean;
  issueEstimationExtended: boolean;
}): EstimateConfig {
  return {
    type: team.issueEstimationType,
    allowZero: team.issueEstimationAllowZero,
    extended: team.issueEstimationExtended,
  };
}

/**
 * Validate an estimate value against a team's configuration.
 * Returns null if valid, or an error message string if invalid.
 */
export function validateEstimate(
  team: {
    key: string;
    issueEstimationType: string;
    issueEstimationAllowZero: boolean;
    issueEstimationExtended: boolean;
  },
  estimate: number,
): string | null {
  if (team.issueEstimationType === "notUsed") {
    return `Team "${team.key}" does not use estimates.`;
  }
  const config = buildEstimateConfig(team);
  const valid = getValidEstimates(config);
  if (!valid.includes(estimate)) {
    const scale = formatEstimateScale(team.issueEstimationType, valid);
    return `Invalid estimate: ${estimate}. Team "${team.key}" uses ${team.issueEstimationType} scale. Valid values: ${scale}`;
  }
  return null;
}

/** Returns the valid numeric estimate values for a team's configuration. */
export function getValidEstimates(config: EstimateConfig): number[] {
  const scale = ESTIMATE_SCALES[config.type];
  if (!scale) {
    return [];
  }
  const values = config.extended ? [...scale.base, ...scale.extended] : [...scale.base];
  return config.allowZero ? [0, ...values] : values;
}

/** Formats valid estimate values for display in error messages. */
export function formatEstimateScale(type: string, values: number[]): string {
  if (type === "tShirt") {
    return values.map((v) => `${v} (${TSHIRT_LABELS[v] ?? v})`).join(" | ");
  }
  return values.join(" | ");
}

File: internal/estimates/estimates.go
package estimates

import (
	"fmt"
	"strings"
)

type Config struct {
	Type      string `json:"type"`
	AllowZero bool   `json:"allowZero"`
	Extended  bool   `json:"extended"`
}

type scale struct {
	base     []int
	extended []int
}

var scales = map[string]scale{
	"exponential": {base: []int{1, 2, 4, 8, 16}, extended: []int{32, 64}},
	"fibonacci":   {base: []int{1, 2, 3, 5, 8, 13}, extended: []int{21, 34}},
	"linear":      {base: []int{1, 2, 3, 4, 5}, extended: []int{6, 7, 8, 9, 10}},
	"tShirt":      {base: []int{1, 2, 3, 4, 5}, extended: []int{6}},
}

var tShirtLabels = map[int]string{
	0: "None", 1: "XS", 2: "S", 3: "M", 4: "L", 5: "XL", 6: "XXL",
}

func BuildConfig(estimationType string, allowZero, extended bool) Config {
	return Config{Type: estimationType, AllowZero: allowZero, Extended: extended}
}

func ValidEstimates(cfg Config) []int {
	s, ok := scales[cfg.Type]
	if !ok {
		return nil
	}
	var values []int
	if cfg.AllowZero {
		values = append(values, 0)
	}
	values = append(values, s.base...)
	if cfg.Extended {
		values = append(values, s.extended...)
	}
	return values
}

func Validate(cfg Config, estimate int) error {
	if cfg.Type == "notUsed" {
		return fmt.Errorf("this team does not use estimates")
	}
	values := ValidEstimates(cfg)
	for _, v := range values {
		if v == estimate {
			return nil
		}
	}
	return fmt.Errorf("invalid estimate %d. Valid values: %s", estimate, FormatScale(cfg.Type, values))
}

func FormatScale(scaleType string, values []int) string {
	parts := make([]string, len(values))
	for i, v := range values {
		if scaleType == "tShirt" {
			label := tShirtLabels[v]
			if label == "" {
				label = fmt.Sprintf("%d", v)
			}
			parts[i] = fmt.Sprintf("%d (%s)", v, label)
		} else {
			parts[i] = fmt.Sprintf("%d", v)
		}
	}
	return strings.Join(parts, " | ")
}
